text/0000-semantic-test-selectors.md

```diff
@@ -0,0 +1,700 @@
+users of assistive technologies and therefore should be invisible to our tests
+too. The corollary: if we want to access a UI element in a test, we do so as if
+we were using a screen reader.
+
+### The rule of least power
+
+> Expressing constraints, relationships and processing instructions in less
+> powerful languages increases the flexibility with which information can be
+> reused: the less powerful the language, the more you can do with the data
+> stored in that language.
+>
+> — <cite>Tim Berners-Lee & Noah Mendelsohn, [The Rule of Least Power]</cite>
+
+Currently, CSS selectors allow us to target elements with arbitrary
+specificity. The more we over-specify in our tests, the less room we leave
+ourselves to reinterpret those tests in the future.
+
+The language of UI tests is already wonderfully constrained: `visit`, `click`,
+`fillIn`, `find`. These helpers, simple at the surface, encode an enormous
+amount of meaning — meaning that is shared between all Ember developers and
+evolves over time. If we could similarly contstrain the means by which we
+select elements, we can open the door for further evolution.
+
+### Cleaner tests
+
+This is perhaps the most visceral motivation. Which test would you rather read?
+
+This one:
+
+```js
+await fillIn('.login-form .field-email', 'alice@example.com');
+await fillIn('.login-form .field-password', 'topsecret');
+await click('.submit-btn');
+```
+
+Or this one:
+
+```js
+await fillIn('Email', 'alice@example.com');
+await fillIn('Password', 'topsecret');
+await click('Log in');
+```
+
+## Detailed design
+
+We will provide semantic variants of `click` and `fillIn`. Their proposed
+behaviours are given below.
+
+### Semantic `click`
+
+```ts
+function click(label: string): Promise<void>
+```
+
+This helper finds a **link** or **button** whose **perceivable text** contains
+the string specified by `label`, or a **switch** whose **perceivable label**
+contains the string specified by `label`.
+
+If it finds exactly one matching element, it invokes the general-purpose
+`click` helper upon it.
+
+If it finds multiple matching elements, it throws:
+
+```
+Found multiple clickable elements with perceivable text '${label}' when calling
+`click('${label}')`. You may want to use a more descriptive label in the test
+and the template.
+```
+
+If it finds no matching element, it throws:
+
+```
+Clickable element with perceivable text '${label}' not found when calling `click('${label}')`.
+```
+
+### Semantic `fillIn`
+
+```ts
+function fillIn(label: string, value: string): Promise<void>
+```
+
+This helper finds an **input** whose **perceivable label** contains the string
+specified by `label`.
+
+If it finds a matching element, it invokes the general-purpose `fillIn` helper
+upon it with `value` as the second argument.
+
+If a label is found without a counterpart form control it throws:
+
+```
+Form control not found for label containing '${label}' when calling
```

Comment:
interesting thing here is that
```
      <label for="hadTitle">hi</label>
      <label for="hadTitle">life</label>
      <input id="hadTitle" title="only-title" />
```

the percieved label is `hilife`

Reply:
`<label>Click me <input type="text"></label>` no for attribute but the input will be labeled with `Click me`